File: split/handlerightbrace.pas
procedure handlerightbrace;
var
  p, q: halfword;
  d: scaled;
  f: integer;
begin
  case curgroup of
    1: unsave;
    0:
    begin
      begin
        if interaction = 3 then;
        printnl(262);
        print(1043);
      end;
      begin
        helpptr := 2;
        helpline[1] := 1044;
        helpline[0] := 1045;
      end;
      error;
    end;
    14, 15, 16: extrarightbrace;
    {1085:}
2: package(0);
    3:
    begin
      adjusttail := 29995;
      package(0);
    end;
    4:
    begin
      endgraf;
      package(0);
    end;
    5:
    begin
      endgraf;
      package(4);
    end;
    {:1085}
{1100:}
11:
    begin
      endgraf;
      q := eqtb[2892].hh.rh;
      mem[q].hh.rh := mem[q].hh.rh + 1;
      d := eqtb[5836].int;
      f := eqtb[5305].int;
      unsave;
      saveptr := saveptr - 1;
      p := vpackage(mem[curlist.headfield].hh.rh, 0, 1, 1073741823);
      popnest;
      if savestack[saveptr + 0].int < 255 then
      begin
        begin
          mem[curlist.tailfield].
            hh.rh := getnode(5);
          curlist.tailfield := mem[curlist.tailfield].hh.rh;
        end;
        mem[curlist.tailfield].hh.b0 := 3;
        mem[curlist.tailfield].hh.b1 := savestack[saveptr + 0].int + 0;
        mem[curlist.tailfield + 3].int := mem[p + 3].int + mem[p + 2].int;
        mem[curlist.tailfield + 4].hh.lh := mem[p + 5].hh.rh;
        mem[curlist.tailfield + 4].hh.rh := q;
        mem[curlist.tailfield + 2].int := d;
        mem[curlist.tailfield + 1].int := f;
      end
      else
      begin
        begin
          mem[curlist.tailfield].hh.rh := getnode(2);
          curlist.tailfield := mem[curlist.tailfield].hh.rh;
        end;
        mem[curlist.tailfield].hh.b0 := 5;
        mem[curlist.tailfield].hh.b1 := 0;
        mem[curlist.tailfield + 1].int := mem[p + 5].hh.rh;
        deleteglueref(q);
      end;
      freenode(p, 7);
      if nestptr = 0 then
        buildpage;
    end;
    8:{1026:}

    begin
      if (curinput.locfield <> 0) or ((curinput.indexfield <> 6) and (curinput.indexfield <> 3)) then{1027:}

      begin
        begin
          if interaction = 3 then;
          printnl(262);
          print(1009);
        end;
        begin
          helpptr := 2;
          helpline[1] := 1010;
          helpline[0] := 1011;
        end;
        error;
        repeat
          gettoken;
        until curinput.locfield = 0;
      end{:1027}
;
      endtokenlist;
      endgraf;
      unsave;
      outputactive := false;
      insertpenalties := 0;
      {1028:}
if eqtb[3933].hh.rh <> 0 then
      begin
        begin
          if interaction = 3 then;
          printnl(262);
          print(1012);
        end;
        printesc(409);
        printint(255);
        begin
          helpptr := 3;
          helpline[2] := 1013;
          helpline[1] := 1014;
          helpline[0] := 1015;
        end;
        boxerror(255);
      end{:1028}
;
      if curlist.tailfield <> curlist.headfield then
      begin
        mem[pagetail].hh.rh :=
          mem[curlist.headfield].hh.rh;
        pagetail := curlist.tailfield;
      end;
      if mem[29998].hh.rh <> 0 then
      begin
        if mem[29999].hh.rh = 0 then
          nest[0].
            tailfield := pagetail;
        mem[pagetail].hh.rh := mem[29999].hh.rh;
        mem[29999].hh.rh := mem[29998].hh.rh;
        mem[29998].hh.rh := 0;
        pagetail := 29998;
      end;
      popnest;
      buildpage;
    end{:1026}
;
    {:1100}
{1118:}
10: builddiscretionary;
    {:1118}
{1132:}
6:
    begin
      backinput;
      curtok := 6710;
      begin
        if interaction = 3 then;
        printnl(262);
        print(625);
      end;
      printesc(898);
      print(626);
      begin
        helpptr := 1;
        helpline[0] := 1124;
      end;
      inserror;
    end;
    {:1132}
{1133:}
7:
    begin
      endgraf;
      unsave;
      alignpeek;
    end;
    {:1133}
{1168:}
12:
    begin
      endgraf;
      unsave;
      saveptr := saveptr - 2;
      p := vpackage(mem[curlist.headfield].hh.rh, savestack[saveptr + 1].int, savestack[saveptr + 0].int, 1073741823);
      popnest;
      begin
        mem[curlist.tailfield].hh.rh := newnoad;
        curlist.tailfield := mem[curlist.tailfield].hh.rh;
      end;
      mem[curlist.tailfield].hh.b0 := 29;
      mem[curlist.tailfield + 1].hh.rh := 2;
      mem[curlist.tailfield + 1].hh.lh := p;
    end;
    {:1168}
{1173:}
13: buildchoices;
    {:1173}
{1186:}
9:
    begin
      unsave;
      saveptr := saveptr - 1;
      mem[savestack[saveptr + 0].int].hh.rh := 3;
      p := finmlist(0);
      mem[savestack[saveptr + 0].int].hh.lh := p;
      if p <> 0 then
        if mem[p].hh.rh = 0 then
          if mem[p].hh.b0 = 16 then
          begin
            if mem[p + 3].hh.rh = 0 then
              if mem[p + 2].hh.rh = 0 then
              begin
                mem[savestack[saveptr + 0].int].hh := mem[p + 1].hh;
                freenode(p, 4);
              end;
          end
          else if mem[p].hh.b0 = 28 then
            if savestack[saveptr + 0].int = curlist.tailfield + 1 then
              if mem[curlist.tailfield].hh.b0 = 16 then{1187:}

              begin
                q :=
                  curlist.headfield;
                while mem[q].hh.rh <> curlist.tailfield do
                  q := mem[q].hh.rh;
                mem[q].hh.rh := p;
                freenode(curlist.tailfield, 4);
                curlist.tailfield := p;
              end{:1187}
;
    end;
    {:1186}
others: confusion(1046)
  end;
end;

File: cpp/src/handlerightbrace.cpp
#include "handlerightbrace.h"
#include "sauvegarde.h"
#include "impression.h"
#include "erreur.h"
#include "endgraf.h"
#include "boite.h"
#include "etat.h"
#include "buildpage.h"
#include "lecture.h"
#include "builddiscretionary.h"
#include "noeud.h"
#include "buildchoices.h"
#include "formule.h"
#include "getnext.h"

void handlerightbrace(char status, Token t, AlignRecordNode* &loop)
{
	switch (curgroup)
	{
		case simple_group: 
			unsave();
			break;
		case bottom_level:
			error("Too many }'s", "You've closed more groups than you opened.\nSuch booboos are generally harmless, so keep going.");
			break;
		case semi_simple_group:
			error("Extra }, or forgotten "+esc("endgroup"), "I've deleted a group-closing symbol because it seems to be\nspurious, as in `$x}$'. But perhaps the } is legitimate and\nyou forgot something else, as in `\\hbox{$x}'. In such cases\nthe way to recover is to insert both the forgotten and the\ndeleted material, e.g., by typing `I$}'.");
			alignstate++;
			break;
		case math_shift_group:
			error("Extra }, or forgotten $", "I've deleted a group-closing symbol because it seems to be\nspurious, as in `$x}$'. But perhaps the } is legitimate and\nyou forgot something else, as in `\\hbox{$x}'. In such cases\nthe way to recover is to insert both the forgotten and the\ndeleted material, e.g., by typing `I$}'.");
			alignstate++;
			break;
		case math_left_group: 
			error("Extra }, or forgotten "+esc("right"), "I've deleted a group-closing symbol because it seems to be\nspurious, as in `$x}$'. But perhaps the } is legitimate and\nyou forgot something else, as in `\\hbox{$x}'. In such cases\nthe way to recover is to insert both the forgotten and the\ndeleted material, e.g., by typing `I$}'.");
			alignstate++;
			break;
		case adjusted_hbox_group:
			adjusttail = adjust_head;
			package(status, box_code, t);
			break;
		case vbox_group:
			endgraf(); [[fallthrough]];
		case hbox_group: 
			package(status, box_code, t);
			break;
		case vtop_group:
			endgraf();
			package(status, vtop_code, t);
			break;
		case insert_group:
		{
			endgraf();
			unsave();
			auto s0 = savestack.back()->int_;
			savestack.pop_back();
			auto p = vpack(head->link, 0, additional);
			popnest();
			if (s0 < 255)
			{
				split_top_skip()->glue_ref_count++;
				tail_append(new InsNode(s0, p->height+p->depth, split_max_depth(), split_top_skip(), floating_penalty(), p->list_ptr));
			}
			else
				tail_append(new AdjustNode(p->list_ptr));
			delete p;
			if (nest.size() == 1)
				buildpage(status);
			break;
		}
		case output_group:
			if (loc || (token_type != output_text && token_type != backed_up))
			{
				error("Unbalanced output routine", "Your sneaky output routine has problematic {'s and/or }'s.\nI can't handle that very well; good luck.");
				do
					t = scanner.get(status);
				while (loc);
			}
			endtokenlist();
			endgraf();
			unsave();
			outputactive = false;
			insertpenalties = 0;
			if (box(255))
				boxerror(255, "Output routine didn't use all of "+esc("box255"), "Your \\output commands should empty \\box255,\ne.g., by saying `\\shipout\\box255'.\nProceed; I'll discard its present contents.");
			if (tail != head)
			{
				pagetail->link = head->link;
				pagetail = tail;
			}
			if (page_head->link)
			{
				if (contrib_head->link == nullptr)
					contrib_tail = pagetail;
				pagetail->link = contrib_head->link;
				contrib_head->link = page_head->link;
				page_head->link = nullptr;
				pagetail = page_head;
			}
			popnest();
			buildpage(status);
			break;
		case disc_group: 
			builddiscretionary();
			scanner.leftBrace(status);
			break;
		case align_group:
			backinput(t);
			inserror(cs_token_flag+frozen_cr, "Missing "+esc("cr")+" inserted", "I'm guessing that you meant to end an alignment here.");
			break;
		case no_align_group:
			endgraf();
			unsave();
			alignpeek(status, loop);
			break;
		case vcenter_group:
		{
			endgraf();
			unsave();
			auto s1 = savestack.back()->int_; //scaled
			savestack.pop_back();
			auto s0 = savestack.back()->int_; //smallnumber
			savestack.pop_back();
			auto p = vpack(head->link, s1, s0);
			popnest();
			auto n = new Noad(vcenter_noad);
			n->nucleus.math_type = sub_box;
			n->nucleus.info = p;
			tail_append(n);
			break;
		}
		case math_choice_group:
			buildchoices();
			scanner.leftBrace(status);
			break;
		case math_group:
		{
			unsave();
			auto &n = *dynamic_cast<NoadContent*>(savestack.back()->index); //NoadContent
			savestack.pop_back();
			n.math_type = sub_mlist;
			auto p = finmlist(nullptr);
			n.info = p;
			if (p && p->link == nullptr)
				switch (p->type)
				{
					case ord_noad:
						if (auto P = dynamic_cast<Noad*>(p); P->subscr.math_type == 0 && P->supscr.math_type == 0)
						{
							n = P->nucleus;
							delete p;
						}
						break;
					case accent_noad:
						if (tail->type == ord_noad && n == dynamic_cast<Noad*>(tail)->nucleus)
						{
							LinkedNode *q;
							followUntilBeforeTarget(head, q, tail); // head -> ... -> q -> tail
							q->link = p; // head -> ... -> q -> p
							delete tail;
							tail = p; // head -> ... -> q -> p=tail
						}
				}
			break;
		}
		default: 
			confusion("rightbrace");
	}
}
